### PR TITLE
Add IdSet and fix IdDict performance problem

### LINKED ISSUE #26043: `IdDict` `length` function is O(N)
Like `Dict`, we should keep a count of the number of elements to provide a fast `length` method.

### PR DESCRIPTION
**Pull request by Keno**
https://github.com/Keno/NewOptimizer.jl requires various sets and dicts. Since those are currently not available at inference time, I'm changing it to use `IdDicts`. However, until now there was no `Set` analogue for `IdDict`, so this PR provides that. Additionally, this fixes #26043, which was causing disastrous performance when using IdDicts rather than Dicts in the new optimizer.

### CODE REVIEW

File: base/abstractdict.jl [hunk 1]
```diff
@@ -659,3 +678,46 @@ get!(d::IdDict{K,V}, @nospecialize(key), @nospecialize(default)) where {K, V} =
...
+pop!(s::IdSet, x) = (pop!(s.dict, x); x)
+pop!(s::IdSet, x, deflt) = x in s ? pop!(s, x) : deflt
+delete!(s::IdSet, x) = (delete!(s.dict, x); s)
+
+sizehint!(s::IdSet, newsz) = (sizehint!(s.dict, newsz); s)
+empty!(s::IdSet) = (empty!(s.dict); s)
+
+filter!(f, d::IdSet) = unsafe_filter!(f, d)
+
+function union!(s::IdSet{T}, itr) where T
```
Comment (JeffBezanson) [inline base/abstractdict.jl]: The existing copy of this method should probably be expanded to `AbstractSet` instead.

Comment (Keno) [reply base/abstractdict.jl]: Will do.

**Comment by Keno** [comment]
CircleCI seems to have started running out of memory, but other than that, this seems good. @JeffBezanson any further review?

**Comment by vtjnash** [comment]
Can we get some local performance numbers? From Travis logs, it looks like this just over doubled the build time and increased the size by ~20-30% for the compiler, but those numbers aren't usually particularly reliable.

**Comment by vtjnash** [comment]
Alright, well, this seems to be a pre-req for #26079 now. So, bump?

**Comment by Keno** [comment]
Was just int he process of rebasing and looking at performance.

**Comment by Keno** [comment]
I see about a 1.3s increase in time to build basecompiler.ji (about 5%) and a 1% increase in basecompiler size. I think that's ok.

**Comment by Keno** [comment]
nanosoldier saw performance regressions in bitset on yesterday's nightly run (https://github.com/JuliaCI/BaseBenchmarkReports/blob/656af6a7f0e5c93b5e0fa03c9e412307c3e46531/daily_2018_2_25/report.md). This seems like the most likely candidate. Investigating.

**Comment by cstjean** [comment]
Any reason why `IdSet` should not be exported?

**Comment by StefanKarpinski** [comment]
As long as we're confident that the API is what we want, I can't see why we wouldn't export it.

**Pull request merged**